packages/massa-web3/src/experimental/basicElements/operationManager.ts

```diff
@@ -0,0 +1,286 @@
+   *
+   * @param operation - The operation to serialize.
+   *
+   * @returns A byte array representing the serialized operation.
+   */
+  static serialize(operation: OperationDetails): Uint8Array {
+    // TODO: check that unsigned.encode is equivalent to varint.encode
+    const components = [
+      unsigned.encode(BigInt(operation.fee)),
+      unsigned.encode(BigInt(operation.expirePeriod)),
+      unsigned.encode(BigInt(operation.type)),
+    ]
+
+    switch (operation.type) {
+      case OperationType.Transaction:
+        operation = operation as TransferOperation
+        components.push(operation.recipientAddress.versionedBytes())
+        components.push(unsigned.encode(BigInt(operation.amount)))
+        break
+      case OperationType.RollBuy:
+      case OperationType.RollSell:
+        operation = operation as RollOperation
+        components.push(unsigned.encode(BigInt(operation.amount)))
+        break
+      case OperationType.ExecuteSmartContractBytecode:
+        operation = operation as ExecuteOperation
+        components.push(unsigned.encode(BigInt(operation.maxGas)))
+        components.push(unsigned.encode(BigInt(operation.coins)))
+        components.push(
+          unsigned.encode(BigInt(operation.contractDataBinary.length))
+        )
+        components.push(operation.contractDataBinary)
+
+        operation.datastore =
+          operation.datastore || new Map<Uint8Array, Uint8Array>()
+        components.push(unsigned.encode(BigInt(operation.datastore.size)))
+
+        // length prefixed key-value pairs encoding
+        for (const [key, value] of operation.datastore) {
+          const keyBytes = Buffer.from(key)
+          const keyLen = unsigned.encode(BigInt(keyBytes.length))
+          const valueBytes = Buffer.from(value)
+          const valueLen = unsigned.encode(BigInt(valueBytes.length))
+          components.push(keyLen, keyBytes, valueLen, valueBytes)
+        }
+        break
+      default:
+        throw new Error('Operation type not supported')
+    }
+
+    return Uint8Array.from(
+      components.flatMap((component) => Array.from(component))
+    )
+  }
+
+  /**
+   * Deserializes an operation according to the Massa protocol.
+   *
+   * @param data - The byte array to deserialize.
+   *
+   * @returns An new instance of OperationDetails representing the deserialized operation.
+   */
+  static deserialize(data: Uint8Array): OperationDetails {
+    let offset = 0
+    const decodeNext = () => {
+      const value = unsigned.decode(data, offset)
+      offset += unsigned.encodingLength(value)
+      return Number(value)
+    }
+
+    let operationDetails: BaseOperation = {
+      fee: decodeNext(),
+      expirePeriod: decodeNext(),
+      type: decodeNext(),
+    }
+
+    switch (operationDetails.type) {
+      case OperationType.Transaction: {
+        const recipientAddress = Address.fromVersionedBytes(
+          data.slice(offset, offset + 33)
```

Comment:
Magic number not clear to me (and not future proof a version could be in more than 1 byte in the future)

Reply:
i didn't find any good solution for that. Except maybe add const attribute to the address object. :shrug:

Reply:
We could, as we do in the node, always return the rest of the data as well as the value (or the next offset)